contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,669 @@
+for using raw block devices, the legacy method of populating the devices and appending it to the device array is sufficient.
+
+# Implementation Plan, Features & Milestones
+
+Phase 1: v1.8
+Feature: Pre-provisioned PVs to precreated devices 
+
+               Milestone 1: API changes
+
+               Milestone 2: Restricted Access 
+              
+               Milestone 3: Changes to the mounter interface as today it is assumed 'file' as the default.
+               
+               Milestone 4: Expose volumeMode to users via kubectl
+               
+               Milestone 5: PV controller binding changes for block devices
+               	       
+	       Milestone 6: Container Runtime changes
+	       
+	       Milestone 7: Initial Plugin changes (GCE, AWS & GlusterFS)
+	       
+	       Milestone 8: Disabling of provisioning where volumeMode == Block is not supported
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+		
+		Milestone 3: Flex volume update
+		
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+# Plugin changes
+## New BlockVolumeMapper interface proposed design
+
+```
+ type BlockVolumeMapper interface {
+       Volume
+       CanBlockMap() error
+       SetUpDevice(podUID types.UID) error
+       SetUpDeviceAt(dir string, podUID types.UID) error
+       GetDevicePath() (string, error)
+       GetGlobalMapPath(spec *Spec) (string, error)
+       GetPodDeviceMapPath() string
+ }
+ 
+ type BlockVolumeUnmapper interface {
+       Volume
+       TearDownDevice() error
+       TearDownDeviceAt(dir string) error
+       GetGlobalUnmapPath(spec *Spec) (string, error)
+       GetPodDeviceUnmapPath() string
+ }
+```
+## Changes for volume mount points
+
+Currently, a volume which has filesystem is mounted to the following two paths on a kubelet node when the volumes is in-use.
+The purpose of those mount points are that Kubernetes manages volume attach/detach status using these mount points and number
+of references to these mount points.
+
+```
+- Global mount path
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/{volumePluginDependentPath}/
+
+- Volume mount path
+/var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{volumeName}/
+```
+
+Even if the volumeMode is "Block", similar scheme is needed. However, the volume which 
+doesn't have filesystem can't be mounted.
+Therefore, instead of volume mount, we use symbolic link to map raw block device.
+Kubelet creates a new symbolic link under the new global map path when volume is attached to a Pod. Number of symbolic links
```

Comment:
"attached to a Pod" means the link is created before, during or after `SetUpDevice` call? Is it related to SetUpDevice at all? (I am trying to detect possible races early)

Reply:
The link is created `after SetUpDevice()`, after some discussion with @msau42.
The role of SetUpDevice() depends on plugin type, w/ attacher or w/o attacher.
- w/ attacher: SetUpDevice() is noop function. just return nil.
- w/o attacher: plugin should attach a volume to a node during this method.

For more detail internal steps;

1. SetUpDevice()
1. Get block device path from GetDevicePath()
1. Create global map path directory and create symlink to block device with name of pod's uuid.
1. Create pod's volume directory and create symlink on pod's volume directory to block device

Reply:
ack, the design is OK and we can discuss details in implementation PRs